packages/repository/src/relations/has-one/has-one.repository.ts

```diff
@@ -0,0 +1,97 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/example-todo
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Getter} from '@loopback/context';
+import {Count, DataObject, Options} from '../../common-types';
+import {Entity} from '../../model';
+import {Filter, Where} from '../../query';
+import {
+  constrainDataObject,
+  constrainFilter,
+  constrainWhere,
+} from '../../repositories/constraint-utils';
+import {EntityCrudRepository} from '../../repositories/repository';
+
+/**
+ * CRUD operations for a target repository of a HasMany relation
+ */
+export interface HasOneRepository<Target extends Entity> {
+  /**
+   * Create a target model instance
+   * @param targetModelData The target model data
+   * @param options Options for the operation
+   * @returns A promise which resolves to the newly created target model instance
+   */
+  create(
+    targetModelData: DataObject<Target>,
+    options?: Options,
+  ): Promise<Target>;
+  /**
+   * Find target model instance
+   * @param Filter A filter object for where, order, limit, etc.
+   * @param options Options for the operation
+   * @returns A promise which resolves with the found target instance(s)
+   */
+  find(filter?: Filter<Target>, options?: Options): Promise<Target[]>;
+
+  /**
+   * Find the only target model instance that belongs to the declaring model.
+   * @param filter Query filter
+   * @param options Options for the operations
+   * @returns A promise of the target object or null if not found.
+   */
+  findOne(filter?: Filter<Target>, options?: Options): Promise<Target | null>;
```

Comment:
Let's replace `find` and `findOne` methods with a single method called `get`.

```ts
get(options?: Options): Promise<Target | undefined>;
```

Since there is always at most one target model instance associated with the source model instance, it feels more natural for me to `get` that single instance than `findOne`. I am also not sure what the `filter` argument would be good for, because I see very little value in filtering a set of at most single item.

On the second thought, maybe we should use `find(option)` instead of `get(options)` to indicate that the method may return `undefined` when the target does not exist, as opposed to throwing an EntityNotFound exception.

I think we should also consider the alternative where `get(options)` actually throws when the target model was not found, because that's what CRUD repository methods `findById`, `patchById`, etc. do.

Reply:
Yeah I thought about not needing `filter` argument too, then thought that someone might want to use the `fields` key in `filter` to get the instance with certain properties (maybe not a good use case). 

>I think we should also consider the alternative where get(options) actually throws when the target model was not found, because that's what CRUD repository methods findById, patchById, etc. do.

These methods throw the `EntityNotFound` error which would require an `id` to be present and in this case we won't be given one for the method. For that reason, I think naming it `get(options)` and returning `undefined` sounds like a solid approach IMO.

Reply:
> Yeah I thought about not needing filter argument too, then thought that someone might want to use the fields key in filter to get the instance with certain properties (maybe not a good use case).

That's a good use case. Can we mark `filter` as accepting `Filter` properties without `where`? See `Exclude<T, U>` in [TypeScript's Advanced Types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#predefined-conditional-types).

> These methods throw the EntityNotFound error which would require an id to be present

I have run into this problem while encountering `belongsTo` relation too. I think it's a clear sign that we need a way how to improve `EntityNotFoundError` to support a where filter in addition to the id value. Such change would be out of scope of this pull request though.

Reply:
> That's a good use case. Can we mark filter as accepting Filter properties without where? See Exclude<T, U> in TypeScript's Advanced Types.

Yes, this approach sounds good to me 👍 and for `EntityNotFound`, we can create a follow-up task for those enhancements.